File: src/Mod/WindLab/windlabexamples/RPS_WavePassageEffect.py
# ***************************************************************************
# *   Copyright (c) 2019 Bernd Hahnebach <bernd@bimstatik.org>              *
# *   Copyright (c) 2020 Sudhanshu Dubey <sudhanshu.thethunder@gmail.com    *
# *                                                                         *
# *   This file is part of the LabRPS CAx development system.              *
# *                                                                         *
# *   This program is free software; you can redistribute it and/or modify  *
# *   it under the terms of the GNU Lesser General Public License (LGPL)    *
# *   as published by the Free Software Foundation; either version 2 of     *
# *   the License, or (at your option) any later version.                   *
# *   for detail see the LICENCE text file.                                 *
# *                                                                         *
# *   This program is distributed in the hope that it will be useful,       *
# *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
# *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
# *   GNU Library General Public License for more details.                  *
# *                                                                         *
# *   You should have received a copy of the GNU Library General Public     *
# *   License along with this program; if not, write to the Free Software   *
# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
# *   USA                                                                   *
# *                                                                         *
# ***************************************************************************

import LabRPS
from LabRPS import Vector as vec

import WindLab
import WindLabObjects

from . import manager
from .manager import init_doc



def get_information():
    return {
        "name": "RPS Wave Passage Effect",
        "windlabfeature": "WavePassageEffect",
    }


def get_explanation(header=""):
    return header + """

To run the example from Python console use:
from windlabexamples.RPS_WavePassageEffect import setup
setup()

"""


def setup(doc=None, showGui=True):
# init LabRPS document. New document will be created if the passed document is null.
    if doc is None:
        doc = init_doc()

    #install the WindLab plugin with all its plugged features
    installResuslt = WindLab.installPlugin("WindLabPlugin")

    # explanation object
    # just keep the following line and change text string in get_explanation method
    manager.add_explanation_obj(doc, get_explanation(manager.get_header(get_information())))

    # create new simulation with default name "Simulation"
    sim = WindLabObjects.makeSimulation(doc)
    
    # check if the simulation is successfully created
    if not sim:
        LabRPS.Console.PrintError("The simulation does not exist.\n")
        return None

    # create new location distribution in the simulation called "Simulation"
    loc = WindLabObjects.makeFeature("SimulationPoints", "Simulation", "Location Distribution",  "Horizontal Distribution")
    
    # check if the created location distribution feature is null
    if not loc:
        LabRPS.Console.PrintError("Error on creating the location distribution.\n")
        return None

    # create new mean wind profile feature in the simulation called "Simulation"
    mean = WindLabObjects.makeFeature("MeanSpeed", "Simulation", "Mean Wind Profile",  "RPS Logarithmic Low")
    
    # check if the created mean wind feature
    if not mean:
        LabRPS.Console.PrintError("The creation of the mean wind profile was not successuful.\n")
        return None

    mean.TerrainRoughness = '0.001266 m' 
    mean.ShearVelocity = '1.76 m/s' 

    # create new frequency distribution in the simulation called "Simulation"
    frequency = WindLabObjects.makeFeature("Frequencies", "Simulation", "Frequency Distribution",  "Double Index Frequency")
    
    # check if the created frequency distribution feature is null
    if not frequency:
        LabRPS.Console.PrintError("Error on creating the frequency distribution.\n")
        return None

    # create new wave passage effect function feature in the simulation called "Simulation"
    wavepassageEff = WindLabObjects.makeFeature("WavePassageEffect", "Simulation", "Wave Passage Effect",  "RPS Wave Passage Effect")
    
    # check if the created mean wind feature
    if not wavepassageEff:
        LabRPS.Console.PrintError("The creation of the wave passage effect was not successuful.\n")

    simPoints = sim.computeLocationCoordinateMatrixP3()

    v1 = vec(simPoints[0][1], simPoints[0][2], simPoints[0][3])
    v2 = vec(simPoints[1][1], simPoints[1][2], simPoints[1][3])
    v3 = vec(simPoints[2][1], simPoints[2][2], simPoints[2][3])

    time = 0.0

    # compute the wave passage effect function between point j and k for all frequencies at time instant of 0s
    wpe12 = sim.computeWavePassageEffectVectorF(v1, v2, time)
    wpe13 = sim.computeWavePassageEffectVectorF(v1, v3, time)
    wpe23 = sim.computeWavePassageEffectVectorF(v2, v3, time)

    # check the computed wave passage effect
    if not wpe13:
        LabRPS.Console.PrintError("The computation of the wave  passage effect has failed.\n")
        return None

    # if we are Gui mode, show the stored result in Alphaplot
    if LabRPS.GuiUp and showGui:
        import WindLabGui
        #show the mean wind speeds in Alphaplot
        WindLabGui.WindLabPyTool.showArray(sim.getSimulationData().numberOfFrequency, 3, wpe12)
        WindLabGui.WindLabPyTool.showArray(sim.getSimulationData().numberOfFrequency, 3, wpe13)
        WindLabGui.WindLabPyTool.showArray(sim.getSimulationData().numberOfFrequency, 3, wpe23)

        import numpy
        import matplotlib.pyplot as plt
        arr1 = numpy.asarray(wpe12)   
        arr2 = numpy.asarray(wpe13)   
        arr3 = numpy.asarray(wpe23)   

        fig, ax = plt.subplots()
        fig.suptitle('Real Part of Wave Passage Effect between Points 1 and 2.')
        ax.plot(arr1[:,0], arr1[:,1], color='r')
        ax.set_xlabel('Frequency (rad/s)')
        ax.set_ylabel('Wave Passage Effect')
        ax.grid(True)
        plt.legend(["Effect 12", "Effect 13", "Effect 23"], loc="lower left")
        ax.set_xscale("log")
        ax.set_yscale("log")
        plt.show()
    doc.recompute()
    return doc

File: src/Mod/WindLabPlugins/WindLabPlugin/RPSWavePassageEffect.cpp
/***************************************************************************
 *   Copyright (c) 2024 Koffi Daniel <kfdani@labrps.com>                   *
 *                                                                         *
 *   This file is part of the LabRPS development system.                   *
 *                                                                         *
 *   This library is free software; you can redistribute it and/or         *
 *   modify it under the terms of the GNU Library General Public           *
 *   License as published by the Free Software Foundation; either          *
 *   version 2 of the License, or (at your option) any later version.      *
 *                                                                         *
 *   This library  is distributed in the hope that it will be useful,      *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU Library General Public License for more details.                  *
 *                                                                         *
 *   You should have received a copy of the GNU Library General Public     *
 *   License along with this library; see the file COPYING.LIB. If not,    *
 *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
 *   Suite 330, Boston, MA  02111-1307, USA                                *
 *                                                                         *
 ***************************************************************************/

#include "RPSWavePassageEffect.h"
#include <Mod/WindLabAPI/App/RPSWindLabFramework.h>
#include <Mod/WindLabTools/App/wavePassageEffect/WavePassageEffect.h>
#include "Widgets/DlgWavePassageEffect.h"
#include <Base/Console.h>
#include <Gui/Control.h>
#include <App/Application.h>
#include <App/Document.h>

using namespace WindLab;
using namespace WindLabAPI;

PROPERTY_SOURCE(WindLab::CRPSWavePassageEffect, WindLabAPI::WindLabFeatureWavePassageEffect)

CRPSWavePassageEffect::CRPSWavePassageEffect()
{
   ADD_PROPERTY_TYPE(ApparentWaveVelocity, (10000.0), "Parameters", App::Prop_None, "The apparent wave velocity.");
   ADD_PROPERTY_TYPE(Coefficient, (5.0), "Parameters", App::Prop_None, "The appropriate coefficient that can obtained from experiments.");
   this->OutputUnit.setValue("Dimensionless");
}


bool CRPSWavePassageEffect::ComputeWavePassageEffectVectorF(const WindLabAPI::WindLabSimulationData& Data, const Base::Vector3d& locationJ, const Base::Vector3d& locationK, const double& dTime, vec& dVarVector, cx_vec& dValVector)
{
    bool returnResult = true;
    
    returnResult = CRPSWindLabFramework::ComputeFrequenciesVectorF(Data, locationJ, dVarVector);
    if (!returnResult)
    {
        return false;
    }

    for (int loop = 0; loop < Data.numberOfFrequency.getValue() && false == Data.isInterruptionRequested.getValue() && true == returnResult; loop++)
    {
        returnResult = ComputeWavePassageEffectValue(Data, locationJ, locationK, dVarVector(loop), dTime, dValVector(loop));
    }

     return returnResult;
}

bool CRPSWavePassageEffect::ComputeWavePassageEffectVectorT(const WindLabAPI::WindLabSimulationData& Data, const Base::Vector3d& locationJ, const Base::Vector3d& locationK, const double& dFrequency, vec& dVarVector, cx_vec& dValVector)
{
     bool returnResult = true;

    for (int loop = 0; loop < Data.numberOfTimeIncrements.getValue() && false == Data.isInterruptionRequested.getValue() && true == returnResult; loop++)
    {
        dVarVector(loop) = Data.minTime.getQuantityValue().getValueAs(Base::Quantity::Second) + loop * Data.timeIncrement.getQuantityValue().getValueAs(Base::Quantity::Second);
        returnResult = ComputeWavePassageEffectValue(Data, locationJ, locationK, dFrequency, dVarVector(loop), dValVector(loop));
    }

     return returnResult;
}

bool CRPSWavePassageEffect::ComputeWavePassageEffectMatrixPP(const WindLabAPI::WindLabSimulationData& Data, const double& dFrequency, const double& dTime, cx_mat& dCoherenceMatrix)
{
     // Local array for location coordinates
    mat dLocCoord(Data.numberOfSpatialPosition.getValue(), 4);
    
    bool returnResult = CRPSWindLabFramework::ComputeLocationCoordinateMatrixP3(Data, dLocCoord);
     if(!returnResult)
    {
       Base::Console().Warning("The computation of the location coordinates matrix has failed.") ;
       return false;
    }
    
    Base::Vector3d locationJ(0, 0, 0);
    Base::Vector3d locationK(0, 0, 0);

    for (int loop1 = 0; loop1 < Data.numberOfSpatialPosition.getValue(); loop1++)
    {
        for (int loop2 = 0; loop2 < Data.numberOfSpatialPosition.getValue(); loop2++)
        {
           locationJ = Base::Vector3d(dLocCoord(loop1, 1), dLocCoord(loop1, 2), dLocCoord(loop1, 3));
           locationK = Base::Vector3d(dLocCoord(loop2, 1), dLocCoord(loop2, 2), dLocCoord(loop2, 3));
           returnResult = ComputeWavePassageEffectValue(Data, locationJ, locationK, dFrequency, dTime, dCoherenceMatrix(loop1, loop2));
        }
    }

    return returnResult;
}

bool CRPSWavePassageEffect::ComputeWavePassageEffectValue(const WindLabAPI::WindLabSimulationData& Data, const Base::Vector3d& locationJ, const Base::Vector3d& locationK, const double& dFrequency, const double& dTime, std::complex<double>& dValue)
{
   bool returnResult = true;

    double MEANj = 0.0;
    double MEANk = 0.0;

    returnResult = CRPSWindLabFramework::ComputeMeanWindSpeedValue(Data, locationJ, dTime, MEANj);
    if (!returnResult) {
        Base::Console().Error("The computation of mean wind speed at location %s has failed.\n");
        return false;
    }

    returnResult = CRPSWindLabFramework::ComputeMeanWindSpeedValue(Data, locationK, dTime, MEANk);
    if (!returnResult) {
        Base::Console().Error("The computation of mean wind speed fails\n");
        return false;
    }

    WindLabTools::WavePassageEffect wavePassageEf;

    dValue = wavePassageEf.computeWavePassageEffect(locationJ, locationK, dFrequency, ApparentWaveVelocity.getQuantityValue().getValueAs(Base::Quantity::MetrePerSecond));

	return true;
}


//Initial setting
bool CRPSWavePassageEffect::OnInitialSetting(const WindLabAPI::WindLabSimulationData& Data)
{
	// the input diolag
    WindLabGui::DlgWavePassageEffectEdit* dlg = new WindLabGui::DlgWavePassageEffectEdit(ApparentWaveVelocity, Coefficient, Data.alongWindSpectrumModel);
	Gui::Control().showDialog(dlg);

    return true;
}